### PR TITLE
[Field] Converting 'negative' finite field elements to negative integers

### PR DESCRIPTION
**Pull request by effectfully**
Now `asInteger (toInteger (-1) :: AField Jubjub.F)` gives `Just (-1)` instead of `52435875175126190479447740508185965837690552500527637822603658699938581184512`.

### CODE REVIEW

File: stack.yaml.lock [hunk 1]
```diff
@@ -52,7 +52,7 @@ packages:
       size: 416
       sha256: c64a9ec9827f140768c01f1d9df1ed9ec9de2d70682a3822061bbe52cf85006c
   original:
-    hackage: quiet-0.2@sha256:60bb3cb8dd4a225557351b6d622147a4e541f1564847e05a81621365a155ad6c,1413
+    hackage: quiet-0.2
```
Comment (effectfully) [inline stack.yaml.lock:55]: It gets the same hash in the `stack.yaml.lock` file anyway.

File: stack.yaml [hunk 2]
```diff
@@ -10,4 +10,4 @@ extra-deps:
 - galois-field-1.0.1
 - bitvec-1.0.2.0
 - vector-algorithms-0.8.0.3
-- quiet-0.2@sha256:60bb3cb8dd4a225557351b6d622147a4e541f1564847e05a81621365a155ad6c,1413
+- quiet-0.2
```
Comment (effectfully) [inline stack.yaml:13]: It gets the same hash in the `stack.yaml.lock` file anyway.

Comment (jakzale) [reply stack.yaml:13]: I thought that specifying shas in `stack.yaml` was more of a security measure.  I was under the impression that lock file might be edited during an update.

File: library/Data/Field.hs [hunk 3]
```diff
@@ -248,18 +234,14 @@ instance IsNegative Rational
...
 
 instance KnownNat p => AsInteger (GF.Prime p) where
-    asInteger = Just . GF.fromP
-
-instance KnownNat p => IsNegative (GF.Prime p) where
-    isNegative f = i > p `Prelude.div` 2 where
-        p = natVal @p Proxy
-        i = GF.fromP f
+    asInteger x
+        | i < p `Prelude.div` 2 = Just i
```
Comment (jakzale) [inline library/Data/Field.hs:238]: Does the edge case of ``i == p `Prelude.div` 2`` matter here. I am not sure.

**Review by jakzale — APPROVED** [review]
lgtm.

File: library/Data/Field.hs (hunk 3 above)
Comment (effectfully) [reply library/Data/Field.hs:238]: I started checking that and realized the whole test is broken and doesn't test what it's supposed to. Will fix.

File: stack.yaml (hunk 2 above)
Comment (effectfully) [reply stack.yaml:13]: But then we'll see that anyway?

**Comment by effectfully** [comment]
Ready for another review. Now we have a test that shows how overflows are handled and the old test now uses `Large Int` instead of `Integer` as that has no chance of overflowing.

File: test/Field/Axioms.hs [hunk 4]
```diff
@@ -102,24 +104,52 @@ test_mulDistributive _ =
...
-        , test_addCommutative            namedType
-        , test_zerAdditiveIdentity       namedType
-        , test_additiveInverse           namedType
-        , test_mulAssociative            namedType
-        , test_mulCommutative            namedType
-        , test_oneMultiplicativeIdentity namedType
-        , test_nonZeroInverse            namedType
-        , test_mulDistributive           namedType
+-- | Test that any 'Int' can be converted to a field element and back, resulting in the same @Int@.
+test_fromToInteger :: forall f proxy. (Field f, AsInteger f) => proxy f -> TestTree
```
Comment (jakzale) [inline test/Field/Axioms.hs:108]: Shouldn't we use `Data.Proxy`?

**Review by jakzale — APPROVED** [review]

File: test/Field/Axioms.hs (hunk 4 above)
Comment (effectfully) [reply test/Field/Axioms.hs:108]: We instantiate that `proxy` with a different thing: `NamedType`. Anyway, it's like 

> Use `void` in positive position for maximum generality
> Use `Void` in negative position to clearly state your demands

(taken from [To Void or to void](https://www.fpcomplete.com/blog/2017/07/to-void-or-to-void)), except the opposite:

> Use `proxy` in negative position for maximum generality
> Use `Proxy` in positive position to clearly state your demands

(as categorically `Proxy` is a terminal object and `Void` is an initial one, hence the duality).

**Pull request merged**